units/bson_to_parquet_backfill.go

```diff
@@ -0,0 +1,183 @@
+	}
+	j.controller = controller
+
+	if j.controller.Timeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, j.controller.Timeout)
+		defer cancel()
+	}
+
+	for i := 0; i < j.controller.Iterations; i++ {
+		if err = j.runIteration(ctx); err != nil {
+			j.AddError(err)
+			return
+		}
+	}
+}
+
+func (j *bsonToParquetBackfillJob) runIteration(ctx context.Context) error {
+	start := time.Now().UTC()
+	collection := j.env.GetDB().Collection(j.controller.Collection)
+
+	// This is really ugly, but unfortunately Mongo does not allow setting
+	// limits on UpdateMany operations so we have to do a BulkWrite with
+	// the batch size number of UpdateOne operations.
+	updates := make([]mongo.WriteModel, j.controller.BatchSize)
+	for i := 0; i < j.controller.BatchSize; i++ {
+		updates[i] = mongo.NewUpdateOneModel().
+			SetFilter(bson.M{"$or": []bson.M{
+				{model.TestResultsMigrationKey: bson.M{"$exists": false}},
+				{bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsVersionKey): bson.M{"$lt": j.controller.Version}},
+				{"$and": []bson.M{
+					{bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsStartedAtKey): bson.M{"$and": []bson.M{
+						{"gt": time.Time{}},
+						{"lt": time.Now().Add(-j.controller.Timeout)},
+					}}},
+					{bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsCompletedAtKey): time.Time{}},
+				}},
+			}}).
+			SetUpdate(bson.M{"$set": bson.M{
+				model.TestResultsMigrationKey: &model.MigrationStats{
+					MigratorID: j.ID(),
+					StartedAt:  start,
+					Version:    j.controller.Version,
+				},
+			}})
+	}
+	bulkWriteResult, err := collection.BulkWrite(ctx, updates)
+	if err != nil {
+		return errors.Wrap(err, "marking test results for migration in temporary collection")
+	}
+	if bulkWriteResult.MatchedCount == 0 {
+		grip.Info(message.Fields{
+			"job_name": bsonToParquetBackfillJobName,
+			"message":  "test results BSON to Parquet backfill complete",
+		})
+		return nil
+	}
+
+	cur, err := collection.Find(
+		ctx,
+		bson.M{
+			bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsMigratorIDKey):  j.ID(),
+			bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsCompletedAtKey): time.Time{},
+		},
+		options.Find().SetLimit(int64(j.controller.BatchSize)),
+	)
+	if err != nil {
+		return errors.Wrap(err, "getting test results from temporary collection")
+	}
+	defer func() {
+		j.AddError(errors.Wrap(cur.Close(ctx), "closing DB cursor"))
+	}()
+
+	var results []model.TestResults
+	if err = cur.All(ctx, &results); err != nil {
+		return errors.Wrap(err, "decoding test results from temporary collection")
+	}
+
+	toUpdate := make([]string, len(results))
+	for i, result := range results {
+		if err = result.DownloadConvertAndWriteParquet(ctx); err != nil {
+			return errors.Wrap(err, "downloading, converting, and writing Parquet test results")
```

Comment:
this will cause the whole batch to fail, ya? Then the other jobs will end up in an "orphaned" state?

Reply:
Yep! But I have a clause in the update filter above to check for test results that find themselves in this state.